common/rfb/CSecurityRSAAES.cxx

```diff
@@ -0,0 +1,444 @@
+ * USA.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#ifndef HAVE_GNUTLS
+#error "This header should not be compiled without HAVE_GNUTLS defined"
+#endif
+
+#include <stdlib.h>
+#ifndef WIN32
+#include <unistd.h>
+#endif
+
+#include <gnutls/gnutls.h>
+#include <gnutls/crypto.h>
+#include <gnutls/x509.h>
+#include <nettle/bignum.h>
+#include <nettle/sha1.h>
+#include <nettle/sha2.h>
+#include <rfb/CSecurityRSAAES.h>
+#include <rfb/CSecurityTLS.h>
+#include <rfb/CConnection.h>
+#include <rfb/LogWriter.h>
+#include <rfb/Exception.h>
+#include <rfb/UserMsgBox.h>
+#include <rdr/AESInStream.h>
+#include <rdr/AESOutStream.h>
+#include <os/os.h>
+
+enum {
+  ReadPublicKey,
+  ReadRandom,
+  ReadHash,
+  ReadSubtype,
+};
+
+const int MinKeyLength = 1024;
+const int MaxKeyLength = 8192;
+
+using namespace rfb;
+
+IntParameter CSecurityRSAAES::RSAKeyLength
+("RSAAESKeyLength", "bits of the key generated for RSA-AES security types",
+  2048, 1024, 8192, ConfViewer);
+
+
+CSecurityRSAAES::CSecurityRSAAES(CConnection* cc, rdr::U32 _secType)
+  : CSecurity(cc), state(ReadPublicKey), secType(_secType),
+    clientKey{0}, clientPublicKey{0}, serverKey{0},
+    serverKeyN(NULL), serverKeyE(NULL),
+    clientKeyLength(RSAKeyLength), clientKeyN(NULL), clientKeyE(NULL),
+    rais(NULL), raos(NULL), rawis(NULL), rawos(NULL)
+{
+  rawis = cc->getInStream();
+  rawos = cc->getOutStream();
+  if (gnutls_global_init() != GNUTLS_E_SUCCESS)
+    throw AuthFailureException("gnutls_global_init failed");
+  try {
+    writePublicKey();
+  } catch (...) {
+    cleanup();
+    throw;
+  }
+  if (secType == secTypeRA2 || secType == secTypeRA2ne)
+    keySize = 16;
+  else
+    keySize = 32;
+}
+
+CSecurityRSAAES::~CSecurityRSAAES()
+{
+  cleanup();
+}
+
+void CSecurityRSAAES::cleanup()
+{
+  gnutls_global_deinit();
+  if (serverKeyN)
+    delete[] serverKeyN;
+  if (serverKeyE)
+    delete[] serverKeyE;
+  if (clientKeyN)
+    delete[] clientKeyN;
+  if (clientKeyE)
+    delete[] clientKeyE;
+  if (clientKey.size)
+    rsa_private_key_clear(&clientKey);
+  if (clientPublicKey.size)
+    rsa_public_key_clear(&clientPublicKey);
+  if (serverKey.size)
+    rsa_public_key_clear(&serverKey);
+  if (secType == secTypeRA2 || secType == secTypeRA256)
+    cc->setStreams(rawis, rawos);
+  if (rais)
+    delete rais;
+  if (raos)
+    delete raos;
+}
+
+bool CSecurityRSAAES::processMsg()
+{
+  switch (state) {
+    case ReadPublicKey:
+      if (readPublicKey()) {
+        verifyServer();
+        writeRandom();
+        state = ReadRandom;
+      }
+      return false;
+    case ReadRandom:
+      if (readRandom()) {
+        setCipher();
+        writeHash();
+        state = ReadHash;
+      }
+      return false;
+    case ReadHash:
+      if (readHash())
+        state = ReadSubtype;
+    case ReadSubtype:
+      if (readSubtype()) {
+        writeCredentials();
+        return true;
+      }
+      return false;
+  }
+  throw AuthFailureException("unreachable");
```

Comment:
An `assert()` is generally more appropriate for protection against coding mistakes.